syft/pkg/cataloger/javascript/parser/yarn/parse.go

```diff
@@ -0,0 +1,313 @@
+		return "", fmt.Errorf("failed to parse version: '%s", target)
+	}
+	return capture[len(capture)-1], nil
+}
+
+func getPackageNameFromResolved(resolution string) (pkgName string) {
+	if matches := yarnPackageURLExp.FindStringSubmatch(resolution); len(matches) >= 2 {
+		return matches[1]
+	}
+	return ""
+}
+
+func parsePattern(target string) (packagename, protocol, version string, err error) {
+	var capture []string
+	var names []string
+
+	if strings.Contains(target, "https://") {
+		capture = yarnPatternHTTPRegexp.FindStringSubmatch(target)
+		protocol = "https"
+		names = yarnPatternHTTPRegexp.SubexpNames()
+	} else {
+		capture = yarnPatternRegexp.FindStringSubmatch(target)
+		names = yarnPatternRegexp.SubexpNames()
+	}
+
+	if len(capture) < 3 {
+		return "", "", "", errors.New("not package format")
+	}
+	for i, group := range names {
+		switch group {
+		case "package":
+			packagename = capture[i]
+		case "protocol":
+			protocol = capture[i]
+		case "version":
+			version = capture[i]
+		}
+	}
+	return
+}
+
+func parsePackagePatterns(target string) (packagename, protocol string, patterns []string, err error) {
+	patternsSplit := strings.Split(target, ", ")
+	packagename, protocol, _, err = parsePattern(patternsSplit[0])
+	if err != nil {
+		return "", "", nil, err
+	}
+
+	var resultPatterns []string
+	for _, pattern := range patternsSplit {
+		_, _, version, _ := parsePattern(pattern)
+		resultPatterns = append(resultPatterns, key.NpmPackageKey(packagename, version))
+	}
+	patterns = resultPatterns
+	return
+}
+
+func validProtocol(protocol string) bool {
+	switch protocol {
+	// example: "jhipster-core@npm:7.3.4":
+	case "npm", "":
+		return true
+	// example: "my-pkg@workspace:."
+	case "workspace":
+		return true
+	// example: "should-type@https://github.com/shouldjs/type.git#1.3.0"
+	case "https":
+		return true
+	}
+	return false
+}
+
+func ignoreProtocol(protocol string) bool {
+	switch protocol {
+	case "patch", "file", "link", "portal", "github", "git", "git+ssh", "git+http", "git+https", "git+file":
+		return true
+	}
+	return false
+}
+
+func handleEmptyLinesAndComments(line string, skipBlock bool) (int, bool) {
+	if len(line) == 0 {
+		return 1, skipBlock
+	}
+
+	if line[0] == '#' || skipBlock {
+		return 0, skipBlock
+	}
+
+	if strings.HasPrefix(line, "__metadata") {
+		return 0, true
+	}
+
+	return 0, skipBlock
+}
+
+func handleLinePrefixes(line string, pkg *PkgRef, scanner *LineScanner) error {
+	switch {
+	case strings.HasPrefix(line, "version"):
+		var err error
+		pkg.Version, err = getVersion(line)
+		return err
+	case strings.HasPrefix(line, "integrity"):
+		var err error
+		pkg.Integrity, err = getIntegrity(line)
+		return err
+	case strings.HasPrefix(line, "resolved"):
+		var err error
+		pkg.Resolved, err = getResolved(line)
+		return err
+	case strings.HasPrefix(line, "dependencies:"):
+		deps := parseDependencies(scanner)
+		pkg.Dependencies = deps
+		return nil
+	default:
+		return nil
+	}
+}
```

Comment:
Wouldn't it be cleaner if using a named return value?

```suggestion

func handleLinePrefixes(line string, pkg *PkgRef, scanner *LineScanner) (err error) {
	switch {
	case strings.HasPrefix(line, "version"):
		pkg.Version, err = getVersion(line)
	case strings.HasPrefix(line, "integrity"):
		pkg.Integrity, err = getIntegrity(line)
	case strings.HasPrefix(line, "resolved"):
		pkg.Resolved, err = getResolved(line)
	case strings.HasPrefix(line, "dependencies:"):
		pkg.Dependencies = parseDependencies(scanner)
	}
	return
}
```